source/common/http/header_map_impl.h

```diff
@@ -256,6 +291,176 @@ class HeaderMapImpl : NonCopyable {
+                    bool coalesced);
+    HeaderEntryImpl(HeaderString&& key, HeaderString&& value, uint32_t index, bool coalesced);
+
+    // HeaderEntry
+    const HeaderString& key() const override { return key_; }
+    void value(absl::string_view value) override;
+    void value(uint64_t value) override;
+    void value(const HeaderEntry& header) override;
+    const HeaderString& value() const override { return value_; }
+    HeaderString& value() override { return value_; }
+
+    HeaderString key_;
+    HeaderString value_;
+    // Used to indicate the logical index of the header entry in the map, and
+    // to determine the order of header entries when iterating over the map.
+    // Note that for iteration, the indices must be different, but not
+    // consecutive (i.e., when removing a header)
+    uint32_t index_;
+    bool coalesced_;
+  };
+
+  using HeaderEagerMapValue = std::list<HeaderEntryImpl>;
+  virtual const HeaderEagerMapValue* findMapElement(const LowerCaseString& key) const;
+  virtual HeaderEagerMapValue* findMapElement(const LowerCaseString& key);
+  HeaderEntryImpl& maybeCreateInline(const LowerCaseString& key);
+  HeaderEntry* getExisting(const LowerCaseString& key);
+  const HeaderEntryImpl* getExistingInline(const LowerCaseString& key) const;
+
+  static uint64_t appendToHeader(HeaderString& header, absl::string_view data,
+                                 absl::string_view delimiter = ",");
+  void updateSize(uint64_t from_size, uint64_t to_size);
+  void addSize(uint64_t size);
+  void subtractSize(uint64_t size);
+
+private:
+  // A map between the header name and a list of HeaderEntryImpl (initialized
+  // to size of a single value, as most headers only appear once). Note that
+  // the values of HeaderEntryImpl.key_ of each entry, will always be equal to the map's key, and
+  // are repeated to return the key and value without creating a new structure.
+  // Also note that we use a list and not a vector variant, because HeaderEntryImpl's copy-c'tor
+  // is disabled.
+  using HeaderEagerMap = absl::flat_hash_map<absl::string_view, HeaderEagerMapValue>;
+  void mapToSortedVector(const HeaderEagerMap& map,
+                         std::vector<const HeaderEntryImpl*>& output_vector) const;
+
+  HeaderEagerMap headers_map_;
+  // Used as a counter of total headers that were added to the map. It is used
+  // to populate the index_ of the HeaderEntryImpl, and does not decrease upon removal
+  // of a header.
+  uint32_t next_header_index_{0};
+  uint32_t all_headers_num_{0};
+
+  // This holds the internal byte size of the HeaderMap.
+  uint64_t cached_byte_size_ = 0;
+};
+
+/**
+ * A HeaderMap implementation that allows mapping (renaming) a header field key
+ * to a different header field key (specifically to support renaming "host" to ":authority").
+ * Adds another map call to the HeaderMap implementation.
+ */
+class RemappedHeaderMapImpl : public HeaderMapImpl {
```

Comment:
It wouldn't be as convenient as the current implementation in many ways, but I suppose we could do this in the codec to avoid this?

Reply:
Haven't thought of that. I assumed this will also need to be supported in the filters, so it should probably be more general.
I'll provide flame graphs of both Request header map and Response header map, to see how much it hurts performance.